### PR TITLE
Fix deffile parser to allow comments before header

### LINKED ISSUE #2211: Comments before "Bootstrap" in definition file not supported in 3.0
Previously an image definition file could start with a comment describing the file (multiple lines starting with #). In the 3.0 version this gives a cryptic error:

```
FATAL:   Unable to create build: unable to get conveyorpacker: invalid build source
```

Removing the initial comment such that `Bootstrap:` are the first characters in the file makes the error go away. I would suggest skipping commented lines such that a brief description of the file can be included on the top as is normal in many programming languages etc

### PR DESCRIPTION
**Pull request by ikaneshiro**
**Description of the Pull Request (PR):**

This PR modifies tokens generated by the deffile parser to include the `%` before the section name. This allows for reliable identification and separate handling of tokens containing section information vs. header information. This moves away from using `IsValidDefinition()` in `/src/pkg/build/build.go` because there isn't a way to know if a file is a __valid__ definition unless it is parsed in its entirety, which is done by `ParseDefinitionFile()` anyway. This also adds more useful error message if no build source is found in a definition that needs one.

**This fixes or addresses the following GitHub issues:**

- Fixes #2211 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](../CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](../CHANGELOG.md) if necessary according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](../CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Review by ArangoGutierrez — DISMISSED** [review]
LGTM

tested with
```
# a
# very
# verbose comment
Bootstrap: library
# a
# very
# verbose comment
From: alpine:3.7
# a
# very
# verbose comment
%runscript
    echo "$@"
# a
# very
# verbose comment
%post
    echo "Hello from inside the container"
    echo "Install additional software here"
```

File: src/pkg/build/types/parser/deffile_test.go [hunk 1]
```diff
@@ -84,6 +84,7 @@ func TestParseDefinitionFile(t *testing.T) {
 		{"Shub", "../../testdata_good/shub/shub", "../../testdata_good/shub/shub.json"},
 		{"Yum", "../../testdata_good/yum/yum", "../../testdata_good/yum/yum.json"},
 		{"Zypper", "../../testdata_good/zypper/zypper", "../../testdata_good/zypper/zypper.json"},
+		{"NoHeader", "../../testdata_good/noheader/noheader", "../../testdata_good/noheader/noheader.json"},
```
Comment (GodloveD) [inline src/pkg/build/types/parser/deffile_test.go:87]: It seems that a def file with no header and another one with comments at the top should be two different tests.

**Review by GodloveD — CHANGES_REQUESTED** [review]
Looks like this still needs a few more 🤗 and 💋'es.

```
05:01 PM 
ubuntu ~/sandbox
$ cat test2.def 

%post
	apt-get update && apt-get install -y curl

05:01 PM 
ubuntu ~/sandbox
$ sudo singularity build --update test test2.def 
WARNING: Authentication token file not found : Only pulls of public images will succeed
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/sylabs/singularity/src/pkg/build/types/parser.doSections(0xc0004e7030, 0xc000095790, 0x1, 0x40)
	/home/david/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:162 +0x421
github.com/sylabs/singularity/src/pkg/build/types/parser.ParseDefinitionFile(0x1201180, 0xc000506210, 0xc000506210, 0x0, 0x0, 0xc0004e9640, 0x140, 0x140, 0x1085520, 0xc000000300, ...)
	/home/david/go/src/github.com/sylabs/singularity/src/pkg/build/types/parser/deffile.go:325 +0x39c
github.com/sylabs/singularity/src/pkg/build.makeDef(0x7ffceb0e9840, 0x9, 0xc0004bec00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/david/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:391 +0x2b5
github.com/sylabs/singularity/src/pkg/build.NewBuild(0x7ffceb0e9840, 0x9, 0x7ffceb0e983b, 0x4, 0x10ff440, 0x3, 0xc0004e0100, 0xc0004e9010, 0x1, 0x1, ...)
	/home/david/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:69 +0x77
github.com/sylabs/singularity/src/cmd/singularity/cli.run(0x19cc3e0, 0xc00019fdd0, 0x2, 0x3)
	/home/david/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/build_linux.go:62 +0x45f
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).execute(0x19cc3e0, 0xc000030070, 0x3, 0x3, 0x19cc3e0, 0xc000030070)
	/home/david/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x19cf5c0, 0x0, 0x68, 0x11f8401)
	/home/david/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).Execute(0x19cf5c0, 0x4, 0x113416c)
	/home/david/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:800 +0x2b
github.com/sylabs/singularity/src/cmd/singularity/cli.ExecuteSingularity()
	/home/david/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/singularity.go:110 +0x91
main.main()
	/home/david/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli.go:16 +0x20

```

Note that the def file above starts with a newline.  Should be fine when used with the `--update` option afaik.

File: src/pkg/build/types/parser/deffile_test.go (hunk 1 above)
Comment (ikaneshiro) [reply src/pkg/build/types/parser/deffile_test.go:87]: Yikes good catch. Ill try to get this panic fixed and more fine grained test cases pushed later tonight

**Comment by ikaneshiro** [comment]
@GodloveD @ArangoGutierrez After some additional thought I have made some additional changes to `IsValidDefinition()`. `IsValidDefinition()` has been changed to essentially be a wrapper for `ParseDefinitionFile()` since the previous method of trying to strip out a header does not always give accurate results. As a consqequence, `canGetHeader()` has also been removed.

**Review by GodloveD — APPROVED** [review]
Sorry it took me a while to get back to this.  lgtm

**Review by bauerm97 — APPROVED** [review]

**Pull request merged**